File: app/src/main/java/me/taosunkist/hello/ui/frgment/progress/SquareProgressBarFragment.java
package me.taosunkist.hello.ui.frgment.progress;

import androidx.fragment.app.Fragment;

import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import android.widget.TextView;

import java.util.Random;

import me.taosunkist.hello.Dimens;
import me.taosunkist.hello.R;

public class SquareProgressBarFragment extends Fragment {
    public SquareProgressBar squareProgressBar;
    private SeekBar progressSeekBar, widthSeekBar;
    private int currentProcess;

    public static SquareProgressBarFragment newInstance() {

        Bundle args = new Bundle();

        SquareProgressBarFragment fragment = new SquareProgressBarFragment();
        fragment.setArguments(args);
        return fragment;
    }

    public static final String TAG = "SquareFragment";

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_square_progress_bar, container, false);

        view.setBackgroundColor(Color.GRAY);

        final TextView progressView = (TextView) view
                .findViewById(R.id.progressDisplay);
        progressView.setText("32%");


        squareProgressBar = view.findViewById(R.id.subi2);
        squareProgressBar.setImage(R.drawable.blenheim_palace);
        squareProgressBar.setColor("#C9C9C9");
        squareProgressBar.setHoloColor(R.color.errorTextColor);
        squareProgressBar.setRoundedCorners(true, Dimens.INSTANCE.getMarginSmall());
        squareProgressBar.setProgress(32);
        squareProgressBar.setWidth(8);
        squareProgressBar.setOnClickListener(view1 -> {

            Random random = new Random();

            // random progress
            setProgressBarProgress(random.nextInt(100), progressView);

            // random width
            int randWidth = random.nextInt(17) + 4;
            widthSeekBar.setProgress(randWidth);
            squareProgressBar.setWidth(randWidth);

            // random colour
            squareProgressBar.setColorRGB(random.nextInt(256), random.nextInt(256), random.nextInt(256));
        });

        progressSeekBar = (SeekBar) view
                .findViewById(R.id.progressSeekBar);
        progressSeekBar
                .setOnSeekBarChangeListener(new OnSeekBarChangeListener() {

                    @Override
                    public void onStopTrackingTouch(SeekBar seekBar) {
                        // nothing to do
                    }

                    @Override
                    public void onStartTrackingTouch(SeekBar seekBar) {
                        // nothing to do
                    }

                    @Override
                    public void onProgressChanged(SeekBar seekBar,
                                                  int progress, boolean fromUser) {
                        setProgressBarProgress(progress, progressView);
                    }
                });

        widthSeekBar = (SeekBar) view.findViewById(R.id.widthSeekBar);
        widthSeekBar.setOnSeekBarChangeListener(new OnSeekBarChangeListener() {

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                // nothing to do
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
                // nothing to do
            }

            @Override
            public void onProgressChanged(SeekBar seekBar, int progress,
                                          boolean fromUser) {
                squareProgressBar.setWidth(progress);
            }
        });
        return view;
    }

    private void setProgressBarProgress(int progress, TextView progressView) {
        squareProgressBar.setProgress(progress);
        progressView.setText(progress + "%");
        progressSeekBar.setProgress(progress);
    }

}

File: app/src/main/java/me/taosunkist/hello/ui/progress/SquareProgressBarFragment.kt
package me.taosunkist.hello.ui.progress

import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.SeekBar
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.tencent.mmkv.MMKV
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.rxkotlin.addTo
import me.taosunkist.hello.R
import me.taosunkist.hello.databinding.FragmentSquareProgressBarBinding
import me.taosunkist.hello.utility.printf
import top.thsunkist.tatame.ui.reusable.GuideTipeLineView
import top.thsunkist.tatame.utilities.Dimens.marginSmall
import java.util.*
import java.util.concurrent.TimeUnit

class SquareProgressBarFragment : Fragment() {
    lateinit var squareProgressBar: SquareProgressBar
    lateinit var progressSeekBar: SeekBar
    lateinit var widthSeekBar: SeekBar
    private val currentProcess = 0

    init {

    }

    lateinit var binding: FragmentSquareProgressBarBinding
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentSquareProgressBarBinding.inflate(inflater, container, false)


        squareProgressBar = binding.subi2
        progressSeekBar = binding.progressSeekBar
        widthSeekBar = binding.widthSeekBar

        val view = binding.root
        Log.i(TAG, requireContext().cacheDir.absolutePath)
        MMKV.initialize(requireContext().cacheDir.absolutePath)
        binding.getExternalCacheDirButton.setOnClickListener { v: View? ->
            Log.i(TAG, "" + MMKV.defaultMMKV().putString("taohui", "sillyb").commit())
            Log.i(TAG, "" + MMKV.defaultMMKV().getString("taohui", "sillyb"))
        }
        view.setBackgroundColor(Color.GRAY)
        val progressView = view
                .findViewById<View>(R.id.progressDisplay) as TextView
        progressView.text = "32%"
        squareProgressBar = view.findViewById(R.id.subi2)
        squareProgressBar.setImage(R.drawable.blenheim_palace)
        squareProgressBar.setColor("#C9C9C9")
        squareProgressBar.setHoloColor(R.color.errorTextColor)
        squareProgressBar.setRoundedCorners(true, marginSmall.toFloat())
        squareProgressBar.setProgress(32)
        squareProgressBar.width = 8
        squareProgressBar.setOnClickListener(View.OnClickListener { view1: View? ->
            val random = Random()

            // random progress
            setProgressBarProgress(random.nextInt(100), progressView)

            // random width
            val randWidth = random.nextInt(17) + 4
            widthSeekBar!!.progress = randWidth
            squareProgressBar.setWidth(randWidth)

            // random colour
            squareProgressBar.setColorRGB(random.nextInt(256), random.nextInt(256), random.nextInt(256))
        })
        progressSeekBar = view
                .findViewById<View>(R.id.progressSeekBar) as SeekBar
        progressSeekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onStopTrackingTouch(seekBar: SeekBar) {
                // nothing to do
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
                // nothing to do
            }

            override fun onProgressChanged(seekBar: SeekBar,
                                           progress: Int, fromUser: Boolean) {
                setProgressBarProgress(progress, progressView)
            }
        })
        widthSeekBar = view.findViewById<View>(R.id.widthSeekBar) as SeekBar
        widthSeekBar!!.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onStopTrackingTouch(seekBar: SeekBar) {
                // nothing to do
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
                // nothing to do
            }

            override fun onProgressChanged(seekBar: SeekBar, progress: Int,
                                           fromUser: Boolean) {
                squareProgressBar.setWidth(progress)
            }
        })

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.lineView.setData(
                Arrays.asList("02-1", "02-1", "02-1", "02-1", "02-1", "02-1", "02-1"),
                Arrays.asList(1.0f, 1.0f, 3.0f, 4.0f, 5.0f, 6.0f, 7.0f))
        binding.guideTipLineView.setOnClickListener {
            binding.guideTipLineView.showGuideType(GuideTipeLineView.GuideType.values()[(0..2).random()])
        }
    }

    private fun setProgressBarProgress(progress: Int, progressView: TextView) {
        squareProgressBar.setProgress(progress)
        progressView.text = "$progress%"
        progressSeekBar.progress = progress
    }

    companion object {
        fun newInstance(): SquareProgressBarFragment {
            val args = Bundle()
            val fragment = SquareProgressBarFragment()
            fragment.arguments = args
            return fragment
        }

        const val TAG = "SquareFragment"
    }
}